src/comomentsEstimators.c

```diff
@@ -0,0 +1,2775 @@
+  
+  UNPROTECT(1);
+  return CM2vec;
+}
+
+SEXP  CM2_CC(SEXP XXc, SEXP rrcoef, SEXP mm11, SEXP mm22, SEXP NN, SEXP PP){
+  /*
+   arguments
+   XXc       : numeric vector with centered observations
+   rrcoef     : double with factor variance
+   mm11      : numeric vector of t(Xc) %*% Xc / NN
+   mm22      : numeric vector of t(Xc^2) %*% Xc^2 / NN
+   NN        : integer, number of observations
+   PP        : integer, number of assets
+   
+   Written by Dries Cornilly
+   */
+  
+  // // declare pointers for the vector
+  double *Xc, *m11, *m22;
+  
+  // use REAL() to access the C array inside the numeric vector passed in from R
+  Xc = REAL(XXc);
+  m11 = REAL(mm11);
+  m22 = REAL(mm22);
+  double rcoef = asReal(rrcoef);
+  double N = asReal(NN);
+  int n = asInteger(NN);
+  int P = asInteger(PP);
+  
+  // allocate and compute the sample variance and covariances
+  SEXP CM2vec = PROTECT(allocVector(REALSXP, 1));
+  double *rCM2vec = REAL(CM2vec);
+  rCM2vec[0] = 0.0;
+  
+  // loop over the unique indices i <= j
+  for (int ii = 0; ii < P; ii++) {
+    int iiP = ii * P;
+    int iiN = ii * n;
+    for (int jj = ii; jj < P; jj++) {
+      int jjP = jj * P;
+      int jjN = jj * n;
+      if (ii == jj) {
+        // element s_ii
+        rCM2vec[0] += (m22[iiP + ii] - m11[iiP + ii] * m11[iiP + ii]) / N;
+      } else {
+        // element s_ij
+        double S31 = 0.0;
+        double S13 = 0.0;
+        for (int tt = 0; tt < n; tt++) {
+          S31 += Xc[iiN + tt] * Xc[iiN + tt] * Xc[iiN + tt] * Xc[jjN + tt];
+          S13 += Xc[iiN + tt] * Xc[jjN + tt] * Xc[jjN + tt] * Xc[jjN + tt];
+        }
+        
+        double temp_ii = S31 / N - m11[jjP + ii] * m11[iiP + ii];
+        double temp_jj = S13 / N - m11[jjP + ii] * m11[jjP + jj];
+        
+        rCM2vec[0] += rcoef * (sqrt(m11[jjP + jj] / m11[iiP + ii]) * temp_ii + 
+          sqrt(m11[iiP + ii] / m11[jjP + jj]) * temp_jj) / N;
+      }
+    } // loop jj
+  } // loop ii
+  
+  UNPROTECT(1);
+  return CM2vec;
+}
+
+
+// // //
+// // Shrinkage and structured estimators for coskewness
+// // //
+
+SEXP  M3_T23(SEXP mmargskews, SEXP PP){
+  /*
+   arguments
+   mmargskews : vector of length PP with marginal skewness values
+   PP        : integer, number of assets
+   
+   Written by Dries Cornilly
+   */
+  
+  // // declare pointers for the vector
+  double *margskews;
+  
+  // use REAL() to access the C array inside the numeric vector passed in from R
+  margskews = REAL(mmargskews);
+  int P = asInteger(PP);
+  
+  // allocate and compute the coskewness matrix
+  SEXP M3 = PROTECT(allocVector(REALSXP, P * (P + 1) * (P + 2) / 6));
+  double *rM3 = REAL(M3);
+  
+  int iter = 0;
+  // loop over the unique indices i <= j <= k
+  for (int ii = 0; ii < P; ii++) {
+    for (int jj = ii; jj < P; jj++) {
+      for (int kk = jj; kk < P; kk++) {
+        // compute coskewness element
+        double elem = 0.0;
+        if ((ii == jj) & (jj == kk)) elem = margskews[ii];
```

Comment:
This should be `&&`, not the bitwise operator `&`.  The latter is likely to cause future bugs.